velox/docs/develop/types.rst

```diff
@@ -146,6 +147,14 @@ Supported range of milliseconds is [0xFFF8000000000000L, 0x7FFFFFFFFFFFF]
 store timezone ID. Supported range of timezone ID is [1, 1680].
 The definition of timezone IDs can be found in ``TimeZoneDatabase.cpp``.
 
+IPADDRESS represents an IPV6 or IPV4 formatted IPV6 address. Its physical
+type is BIGINT. The format that the address is stored in is defined as part of `(RFC 4291#section-2.5.5.2) <https://datatracker.ietf.org/doc/html/rfc4291.html#section-2.5.5.2>`_
+As Velox is run on Little Endian systems and the standard is network byte(Big Endian)
+order, we reverse the bytes to allow for masking and other bit operations
+used in IPADDRESS/IPPREFIX related functions. This type can be used to
```

Comment:
Is this an orderable type? If so, does OrderBy produce correct results if it treats values of this type the same as values of BIGINT type? If not, we may have a problem similar to https://github.com/facebookincubator/velox/issues/10338

CC: @pedroerp @bikramSingh91

Reply:
@mbasmanova 
I believe that it is working somewhat correctly. To confirm I do have an E2E test that was added for my complete PR of IPADDRESS/IPPREFIX/functions. 
https://github.com/prestodb/presto/pull/23282

I believe that the reason for this is due to the 1 to 1 mapping of IPADDRESS to BIGINT and how they are stored.

For example the ipaddress 
`1.1.1.1` in hex is `00000000000000000000ffff01010101` and which has a binary value of `281470698586369`

`1.1.1.2` in hex is `00000000000000000000ffff01010102` which has a binary value of `281470698586370` which is one larger. 

`1.1.2.1` in hex is `00000000000000000000ffff01010201` which is `281470698586625` which is larger than both `1.1.1.1` and `1.1.1.2` as it should be. 

Similar with IPV6. 

That being said that does point out two issues. One that I don't think presto even takes care of. 
1. When we sort a mixture of IPV4 and IPV6 there will be a split between IPV6 and IPV4 addresses at `::FFFF:FFFF:FFFF` This one should exist in presto. 
So the ordering will be for example

```
::FFFE:FFFF:FFFF
0.0.0.0
....
255.255.255.255
::1:0000:0000:0000
```

2. The other problem being that any IPV6 IPADDRESS with the MSB set will be placed at the start instead of at the end.  This one shouldn't exist in presto as its doing a compareUnsigned. 
For example 
````
FFFF:FFFF:FFFF:FFFF:FFFF:FFFF:FFFF:FFFF
::
::1
````

Reply:
Thank you for looking into this. I'll need more time to understand.

I see that compare in Presto for IPADDRESS is not the same as for BIGINT (or maybe they just look different). This suggest that IPADDRESS type is affected by https://github.com/facebookincubator/velox/issues/10338 and therefore we won't be able to use it (without causing correctness issues) until that issue is resolved or we find a way to block code paths that may be affected.

```
    @Override
    public int compareTo(Block leftBlock, int leftPosition, Block rightBlock, int rightPosition)
    {
        int compare = Long.compareUnsigned(reverseBytes(leftBlock.getLong(leftPosition, 0)), reverseBytes(rightBlock.getLong(rightPosition, 0)));
        if (compare != 0) {
            return compare;
        }
        return Long.compareUnsigned(reverseBytes(leftBlock.getLong(leftPosition, SIZE_OF_LONG)), reverseBytes(rightBlock.getLong(rightPosition, SIZE_OF_LONG)));
    }
```

Reply:
No problem. Simplest fix would be to switch to varbinary but that is not very efficient. Other option would be to override the operator similar to how we override cast? Not sure if this is possible.

Reply:
> Not sure if this is possible

Not yet. @bikramSingh91 has been looking into this as part of https://github.com/facebookincubator/velox/issues/10338

Reply:
@mohsaka One option might be to proceed to land this change and continue landing functions that use that type, but document that it is not "ready to be used" and add a plan validator to Presto to reject plans with that type when running native engine (we have that for TS_w_TS type).

Reply:
@mohsaka We should add the plan validator with a config like it was done for timestamp with timezone in Presto coordiantor first before merging this change in velox. 
Example:
https://github.com/prestodb/presto/pull/23200
https://github.com/prestodb/presto/pull/23374
It will allow us to test it, but disable for production or correctness critical use cases. If we merge this and worker start executing queries with IPAddress type unconditonally, we will face correctness problems.

Reply:
@amitkdutta Thank you for the pointers. Will work on adding the plan changes before merging in these changes.